core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt

```diff
@@ -34,10 +34,9 @@ class UnitMovementAlgorithms(val unit: MapUnit) {
         val toOwner = to.getOwner()
         val extraCost = if (
             toOwner != null &&
-            to.isLand &&
-            toOwner.hasActiveGreatWall &&
+            toOwner.hasActiveEnemyMovementPenalty &&
             civInfo.isAtWarWith(toOwner)
-        ) 1f else 0f
+        ) toOwner.getEnemyMovementPenalty(unit) else 0f
```

Comment:
This calls an additional function from `getMovementCostBetweenAdjacentTiles`, which (as per the comment above the function) should be as time-optimal as possible. Have you measured the impact this change has on a game, both with and without a modded wonder with the new unique?

Reply:
I haven't measured it yet but I kept the condition of this `if` statement as close to the original as I could which should hopefully keep performance similar to before without jeopardizing modability. `CivilizationInfo.hasActiveEnemyMovementPenalty` is set once each turn when `GameInfo.setTransients()` is called (or if a civ builds a building that affects `CivilizationInfo.hasActiveEnemyMovementPenalty`). This boolean variable is used in the condition of this `if` statement 2 lines above so we can know at a glance if we should even bother calculating if there is a movement penalty. The idea being that 1 calculation per turn to set a boolean variable for each `civInfo` then checking this boolean variable every time we try to find a path should keep everything as time efficient as possible.

Reply:
I agree that the case where the condition of the `if` is false is well-optimized, but I'm a bit more worried about the case where it is true. In this case we call another function, which then calls `getMatchingUniques`, which feels like a very costly function

Reply:
That's a good point. I've changed it so that there's a new variable that stores the movement penalty uniques in a `Sequence` so we only have to iterate over the sequence rather than call `getMatchingUniques` for every tile that a unit tries to travel to. This cache is updated every time the game updates movement penalty uniques which is at most twice per turn in a base Unciv game (start of turn then if the civ builds the Great Wall) rather than calling `getMatchingUniques` several hundred times each AI turn potentially.